V20 alpha 3: Cannot invoke "io.swagger.v3.oas.models.media.Schema.getName()" because "schema" is null

### Description of the bug / feature

A project with an endpoint like this does not compile
```
@Endpoint
@AnonymousAllowed
public class MyEnd {

    public void hello(FakePageable p) {

    }
}
```

The source of FakePageable is here https://github.com/Artur-/test-schema/blob/master/src/main/java/com/example/application/FakePageable.java along with the whole test project

### Minimal reproducible example
1. git clone https://github.com/Artur-/test-schema.git
2. mvn

### Expected behavior
Project starts, as it does with V20 alpha 2

### Actual behavior
```
Caused by: java.lang.NullPointerException: Cannot invoke "io.swagger.v3.oas.models.media.Schema.getName()" because "schema" is null
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.collectUsedTypesFromSchema(OpenApiObjectGenerator.java:487)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.collectUsedTypesFromSchema(OpenApiObjectGenerator.java:503)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.lambda$collectUsedTypesFromSchema$21(OpenApiObjectGenerator.java:517)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.collectUsedTypesFromSchema(OpenApiObjectGenerator.java:516)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.generatedRelatedSchemas(OpenApiObjectGenerator.java:758)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.parseNonEndpointClassAsSchema(OpenApiObjectGenerator.java:452)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.createSchemasFromQualifiedNameAndType(OpenApiObjectGenerator.java:476)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.init(OpenApiObjectGenerator.java:228)
	at com.vaadin.flow.server.connect.generator.OpenApiObjectGenerator.generateOpenApi(OpenApiObjectGenerator.java:181)
	at com.vaadin.flow.server.connect.generator.OpenApiSpecGenerator.generateOpenApiSpec(OpenApiSpecGenerator.java:80)
	at com.vaadin.flow.server.connect.generator.OpenApiSpecGenerator.generateOpenApiSpec(OpenApiSpecGenerator.java:112)
	at com.vaadin.flow.server.frontend.fusion.TaskGenerateOpenApiImpl.execute(TaskGenerateOpenApiImpl.java:64)
	at com.vaadin.flow.server.frontend.NodeTasks.execute(NodeTasks.java:698)
	at com.vaadin.flow.server.startup.DevModeInitializer.runNodeTasks(DevModeInitializer.java:487)
	at com.vaadin.flow.server.startup.DevModeInitializer.lambda$initDevModeHandler$0(DevModeInitializer.java:377)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1800)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1792)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

### Versions:
    - Vaadin / Flow version: 20 alpha 3

--- a/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
+++ b/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
@@ -168,7 +168,7 @@ class SchemaResolver {
         if (isTypeOf(type, Collection.class)) {
             return true;
         }
-        if (Iterable.class.getName().equals(type.asReferenceType().getQualifiedName())) {
+        if (isType(type, Iterable.class)) {
             return true;
         }
         return false;
@@ -194,6 +194,16 @@ class SchemaResolver {
         }
     }
 
+    /**
+     * Checks if the given type is an instance of the given class
+     */
+    private boolean isType(ResolvedType type, Class clazz) {
+        return clazz.getName().equals(type.asReferenceType().getQualifiedName());
+    }
+
+    /**
+     * Checks if the given type can be cast to one of the given classes
+     */
     private boolean isTypeOf(ResolvedType type, Class... clazz) {
         if (!type.isReferenceType()) {
             return false;

--- a/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
+++ b/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
@@ -197,14 +197,17 @@ class SchemaResolver {
     /**
      * Checks if the given type is an instance of the given class
      */
-    private boolean isType(ResolvedType type, Class clazz) {
+    private boolean isType(ResolvedType type, Class<?> clazz) {
+        if (!type.isReferenceType()) {
+            return false;
+        }
         return clazz.getName().equals(type.asReferenceType().getQualifiedName());
     }
 
     /**
      * Checks if the given type can be cast to one of the given classes
      */
-    private boolean isTypeOf(ResolvedType type, Class... clazz) {
+    private boolean isTypeOf(ResolvedType type, Class<?>... clazz) {
         if (!type.isReferenceType()) {
             return false;
         }

--- a/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
+++ b/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
@@ -161,17 +161,7 @@ class SchemaResolver {
     }
 
     private boolean isCollectionType(ResolvedType type) {
-        if (type.isPrimitive()) {
-            return false;
-        }
-
-        if (isTypeOf(type, Collection.class)) {
-            return true;
-        }
-        if (isType(type, Iterable.class)) {
-            return true;
-        }
-        return false;
+        return !type.isPrimitive() && (isTypeOf(type, Collection.class) || isType(type, Iterable.class));
     }
 
     private boolean isMapType(ResolvedType type) {
@@ -195,7 +185,13 @@ class SchemaResolver {
     }
 
     /**
-     * Checks if the given type is an instance of the given class
+     * Checks if the given type refers to the given class.
+     *
+     * @param type
+     *            type type to check
+     * @param clazz
+     *            the class to match with
+     * @return true if the type is referring to the given class, false otherwise
      */
     private boolean isType(ResolvedType type, Class<?> clazz) {
         if (!type.isReferenceType()) {
@@ -205,7 +201,13 @@ class SchemaResolver {
     }
 
     /**
-     * Checks if the given type can be cast to one of the given classes
+     * Checks if the given type can be cast to one of the given classes.
+     *
+     * @param type
+     *            type type to check
+     * @param clazz
+     *            the classes to match with
+     * @return true if the type can be cast to one of the given classes, false otherwise
      */
     private boolean isTypeOf(ResolvedType type, Class<?>... clazz) {
         if (!type.isReferenceType()) {

--- a/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
+++ b/fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/generator/SchemaResolver.java
@@ -161,7 +161,17 @@ class SchemaResolver {
     }
 
     private boolean isCollectionType(ResolvedType type) {
-        return !type.isPrimitive() && (isTypeOf(type, Collection.class) || isTypeOf(type, Iterable.class));
+        if (type.isPrimitive()) {
+            return false;
+        }
+
+        if (isTypeOf(type, Collection.class)) {
+            return true;
+        }
+        if (Iterable.class.getName().equals(type.asReferenceType().getQualifiedName())) {
+            return true;
+        }
+        return false;
     }
 
     private boolean isMapType(ResolvedType type) {

--- a/fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/generator/SchemaResolverTest.java
+++ b/fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/generator/SchemaResolverTest.java
@@ -16,9 +16,11 @@
 package com.vaadin.flow.server.connect.generator;
 
 import java.time.Instant;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Date;
+import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
@@ -258,6 +260,16 @@ public class SchemaResolverTest {
         Assert.assertEquals(1, schemaResolver.getFoundTypes().size());
     }
 
+    @Test
+    public void should_ReturnBeanSchema_When_GivenTypeIsABeanImplementingIterable() {
+        ResolvedType resolvedType = mockReferencedTypeOf(TestIterableBean.class);
+        Schema schema = schemaResolver.parseResolvedTypeToSchema(resolvedType);
+        Assert.assertTrue(schema instanceof ObjectSchema);
+        String beanRef = schemaResolver.getFullQualifiedNameRef(TestIterableBean.class.getCanonicalName());
+        Assert.assertEquals(beanRef, schema.get$ref());
+        Assert.assertEquals(1, schemaResolver.getFoundTypes().size());
+    }
+
     private ResolvedType mockReferencedTypeOf(Class clazz) {
         ResolvedType resolvedType = mock(ResolvedType.class);
         ResolvedReferenceType resolvedReferenceType = mock(
@@ -266,8 +278,15 @@ public class SchemaResolverTest {
         when(resolvedType.isPrimitive()).thenReturn(false);
         when(resolvedType.isReferenceType()).thenReturn(true);
         when(resolvedType.asReferenceType()).thenReturn(resolvedReferenceType);
-        when(resolvedReferenceType.getQualifiedName())
-                .thenReturn(clazz.getCanonicalName());
+        when(resolvedReferenceType.getQualifiedName()).thenReturn(clazz.getCanonicalName());
+        List<ResolvedReferenceType> ancestors = new ArrayList<>();
+        for (Class<?> c : clazz.getInterfaces()) {
+            ancestors.add(mockReferencedTypeOf(c).asReferenceType());
+        }
+        if (clazz.getSuperclass() != null) {
+            ancestors.add(mockReferencedTypeOf(clazz.getSuperclass()).asReferenceType());
+        }
+        when(resolvedReferenceType.getAllAncestors()).thenReturn(ancestors);
         return resolvedType;
     }
 
@@ -283,4 +302,13 @@ public class SchemaResolverTest {
     private static class TestBean {
         String foo;
     }
+
+    private static class TestIterableBean implements Iterable<String> {
+        String foo;
+        @Override
+        public Iterator<String> iterator() {
+            return Collections.emptyIterator();
+        }
+    }
+
 }

